Update Elixir and Phoenix and all deps

lleger opened the pull request:
- [x] Look over your code one last time.
- [x] Test all functionality you expect to be QA'd. Don’t forget to QA user interface changes in the browser.
- [x] Describe what changes this PR includes.
- [x] Post screenshots for design review if applicable.
- [x] Fill in QA Notes listing clear, ordered steps for testers.

## Description

Also updates generated files from new Phoenix project. I updated this
manually comparing a fresh Phoenix 1.6.11. I also used the [PhoenixDiff]
project to double check.

[phoenixdiff]: https://www.phoenixdiff.org/?source=1.5.6&target=1.6.11

Closes [sc-20388].

## Screenshots

N/A

## QA Notes

PR Author Browser Info: N/A

```
**QA Tester Browser Info:**

## Environment Setup

- [ ] `asdf install`
- [ ] `mix deps.get`

## Testing Steps

- [ ] Tests should pass
- [ ] App should boot

## Notes

- [ ] [Tester should add any comments, questions, or concerns that pop up during QA]
```

lleger committed:
Update Elixir and Phoenix and all deps

Also updates generated files from new Phoenix project. I updated this
manually comparing a fresh Phoenix 1.6.11. I also used the [PhoenixDiff]
project to double check.

[phoenixdiff]: https://www.phoenixdiff.org/?source=1.5.6&target=1.6.11

lleger commented on Dockerfile:
```diff
@@ -23,9 +23,11 @@ ENV MIX_ENV=prod
-COPY config config
-COPY mix.* ./
-RUN mix do deps.get --only=$MIX_ENV, deps.compile
+COPY mix.exs mix.lock ./
```
The changes here are mostly due to how Docker caches. Each line creates a new layer. Based on what needs to be recompiled, separating some steps and copying at different points helps maximize the Docker cache, which speeds up the build. 

These are picked up from the new Phoenix default Dockerfile.

lleger commented on assets/package.json:
```diff
@@ -2,17 +2,13 @@
-    "deploy": "webpack --mode production",
-    "watch": "webpack --mode development --watch"
-  },
+  "scripts": {},
```
@ngscheurich Any thoughts on how to write a script here that mirrors `mix lint` I added to combine all the JS linters/formatters we have in CI into one command to run locally?

ngscheurich replied:
@lleger Yep! I'm going to add a commit for this.

lleger commented on lib/phoenix_starter/users/user_notifier.ex:
```diff
@@ -2,68 +2,65 @@ defmodule PhoenixStarter.Users.UserNotifier do
-  def deliver_confirmation_instructions(user, url, true) do
-    %{email: "confirmation_instructions", user_id: user.id, url: url}
-    |> UserEmailWorker.new()
-    |> Oban.insert()
```
The Swoosh README (written by Jose) recommends not doing this unless you have a business case for it, so by default I'm removing this code. It can be added later if the specific app requires it.

lleger commented on lib/phoenix_starter_web/channels/user_socket.ex:
```diff
@@ -1,35 +0,0 @@
-defmodule PhoenixStarterWeb.UserSocket do
```
These are just not generated by default anymore and instead are in a generator.

lleger commented on .github/workflows/ci.yml:
```diff
@@ -88,10 +88,6 @@ jobs:
         working-directory: ./assets
       - run: npx eslint js --ext .js,.jsx,.ts,.tsx
         working-directory: ./assets
-      - run: npm run deploy
```
This was running web pack, which we no longer need to do.

lleger replied:
@ngscheurich Is there a lint we can run for Typescript to ensure our types pass? I think this deploy step was added to get the TS checks.

ngscheurich replied:
Unfortunately, esbuild does not do any type-checking so we'll still need to depend on the `typescript` package and run `npx tsc --noEmit` in CI.

lleger commented on .github/workflows/ci.yml:
```diff
@@ -174,9 +170,7 @@ jobs:
       - name: Create PLTs
         if: steps.plt-cache.outputs.cache-hit != 'true'
-        run: |
-          mkdir -p priv/plts
```
No longer needed, Dialyxir does this automatically now.

lleger replied:
https://github.com/jeremyjh/dialyxir/pull/368

lleger commented on mix.exs:
```diff
@@ -69,17 +71,24 @@ defmodule PhoenixStarter.MixProject do
     [
       ignore_warnings: ".dialyzer_ignore.exs",
       plt_file: {:no_warn, "priv/plts/dialyzer.plt"},
-      flags: [:error_handling, :race_conditions, :underspecs, :unknown],
```
race_conditions dropped in OTP 25

lleger replied:
https://github.com/erlang/otp/commit/74c65fbb588b98ee24df9f7302a43552178dfac2

ngscheurich commented on lib/phoenix_starter_web/live/user_settings/index.html.heex:
```diff
@@ -1,26 +1,25 @@
-  <%= live_patch "Update profile", to: Routes.user_settings_path(@socket, :profile) %>
-  <%= live_patch "Update email", to: Routes.user_settings_path(@socket, :email) %>
-  <%= live_patch "Update password", to: Routes.user_settings_path(@socket, :password) %>
+  <%= live_patch("Update profile", to: Routes.user_settings_path(@socket, :profile)) %>
```
I guess the `Phoenix.LiveView.HTMLFormatter` added these parens all over the place? I don't like it but I suppose we should abide by the formatter.

ngscheurich reviewed (approved):
Looks great.

ngscheurich committed:
Skip type checking all .d.ts files

ngscheurich committed:
Add NPM scripts

ngscheurich committed:
fixup! Add NPM scripts

ngscheurich committed:
Ignore config files

ngscheurich commented on .github/workflows/ci.yml:
```diff
@@ -82,16 +82,7 @@ jobs:
-        env:
-          NODE_ENV: production
-        working-directory: ./assets
+      - run: npm run ci
```
@lleger Replaced the individual run steps with a single NPM script.

ngscheurich commented on assets/package.json:
```diff
@@ -3,16 +3,23 @@
   "scripts": {
-    "deploy": "webpack --mode production",
-    "watch": "webpack --mode development --watch"
+    "ci": "npm run lint && npm run format:check && npm run type-check",
+    "format": "prettier --write .",
+    "format:check": "prettier --check .",
+    "lint": "npm run lint:css && npm run lint:js",
+    "lint:css": "stylelint css/**/*.css",
+    "lint:css:fix": "stylelint css/**/*.css --fix",
+    "lint:fix": "npm run lint:css:fix && npm run lint:js:fix",
+    "lint:js": "eslint . --ext .js,.jsx,.ts,.tsx",
+    "lint:js:fix": "eslint . --ext .js,.jsx,.ts,.tsx --fix",
+    "type-check": "tsc --noEmit"
```
@lleger How do these scripts look to you?

ngscheurich commented on assets/tsconfig.json:
```diff
@@ -6,6 +6,7 @@
     "module": "es6",
     "moduleResolution": "node",
     "noImplicitAny": true,
+    "skipLibCheck": true,
```
This is standard in TypeScript projects initialized with `tsc --init` and now we don't have to remember to include it as a CLI arg.

ngscheurich commented:
@lleger I added some NPM scripts for linting, formatting, and type checking. These surfaced some TypeScript errors which I'll have a look at tomorrow.

ngscheurich committed:
fixup! fixup! Add NPM scripts

ngscheurich committed:
Update TypeScript configuration

This aligns us with current best practices documented at
https://www.typescriptlang.org/tsconfig.

ngscheurich commented on assets/tsconfig.json:
```diff
@@ -1,12 +1,21 @@
-    "esModuleInterop": true,
+    /* Language and Environment */
+    "target": "ES6",
+    "lib": ["DOM", "ES2017"],
     "jsx": "react",
+
+    /* Modules */
     "module": "es6",
     "moduleResolution": "node",
-    "noImplicitAny": true,
+
+    /* Interop Constraints */
+    "esModuleInterop": true,
+
+    /* Type Checking */
     "strict": true,
-    "target": "es5"
+
+    /* Completeness */
+    "skipLibCheck": true
```
This aligns us with current best practices documented at https://www.typescriptlang.org/tsconfig.
